File: 4/4-1.js
// BEGIN boring part
const fs = require("fs")
const raw = fs.readFileSync(`${__dirname}/input.txt`, "utf8")
const lines = raw.split("\n")
const input = lines.map(line => {
  const [time, state] = line
    .split(/[|]|Guard|falls|wakes/)
    .map(v => v.trim().replace(/\[|\]/g, ""))
  parsedTime = new Date(time)
  timestamp = Math.round(new Date(time).getTime() / 1000)
  minute = parsedTime.getMinutes()
  return { time, parsedTime, state, timestamp, minute }
})
// END boring part
const range = (start, end) =>
  Array.from({ length: end - start }, (v, k) => k + start)
const getUpdatedSleptMinutes = (data, range) => {
  const newData = []
  for (let i = range[0]; i < range[range.length - 1]; i++) {
    newData[i] = isNaN(data[i]) ? 1 : 2
	}
  return newData
}


const guards = new Map()
sortedEvents = input.sort((a, b) => a.timestamp - b.timestamp)

let sleepTime = -1
let currentguard = ""

for (e of sortedEvents) {
  const currentState = e.state.substring(0, 1)
  switch (currentState) {
    case "#": //shift begins
      currentguard = e.state.substring(0, e.state.indexOf(" "))
      if (!guards.has[currentguard]) {
        guards.set(currentguard, [])
      }
      break
    case "a": //fell asleep
      sleepTime = e.minute
      break
    case "u": //up
      upTime = e.minute
      const data = guards.get(currentguard)
      guards.set(
        currentguard,
        getUpdatedSleptMinutes(data, range(sleepTime, upTime))
      )
      break
    default:
      break
  }
}
guards.forEach(function(value, key) {
  console.log(key + " = " + JSON.stringify(value))
})

File: 4/4-1.go
package main

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

type Moment struct {
	Timestamp int64  `json:"time"`
	Minute    int    `json:"minute"`
	Msg       string `json:"msg"`
}

type Guard struct {
	Id int `json:"id"`
}

func main() {
	absPath, _ := filepath.Abs("./input.txt")
	b, err := ioutil.ReadFile(absPath) // just pass the file name
	layout := "2006-01-02 15:04"
	if err != nil {
		fmt.Print(err)
	}

	str := string(b) // convert content to a 'string'

	split := strings.Split(str, "\r")

	moments := []Moment{}
	for _, line := range split {

		s := strings.Split(line, "]")
		t, msg := s[0], s[1]
		shiftTime, err := time.Parse(layout, strings.TrimSpace(t[strings.Index(t, "[")+1:len(t)]))
		_, min, _ := shiftTime.Clock()
		moment := new(Moment)
		moment.Minute = min
		moment.Timestamp = shiftTime.UnixNano() / int64(time.Millisecond)
		moment.Msg = strings.TrimSpace(msg)
		moments = append(moments, *moment)
		if err != nil {
			fmt.Println(err)
		}
	}
	sort.Slice(moments, func(i, j int) bool {
		return moments[i].Timestamp < moments[j].Timestamp
	})

	for _, line := range moments {

		if strings.Contains(line.Msg, "begins shift") {
			guards[line.Msg] = line.Msg
			fmt.Println(line.Msg)
		}
	}
}
